src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+pub unsafe fn test_checkerboard(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestCheckerboard");
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter = u64::try_from(
+        (half_count * 2)
+            .checked_mul(64)
+            .context("expected_iter overflowed")?,
+    )
+    .context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    let mut checker_board: usize = 0;
+    write_bytes(&mut checker_board, 0x55, 1);
+
+    for i in 0..64 {
+        let val = if i % 2 == 0 {
+            checker_board
+        } else {
+            !checker_board
+        };
+        for j in 0..half_count {
+            timeout_checker.check()?;
+            let val = if j % 2 == 0 { val } else { !val };
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write to all bytes with the value i in an alternating pattern
+/// Read and compare the two halves of the memory region
+/// This procedure is repeated 256 times, with i corresponding to the iteration number 0-255
+pub unsafe fn test_block_seq(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestBlockSeq");
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter = u64::try_from(
+        (half_count * 2)
+            .checked_mul(256)
+            .context("expected_iter overflowed")?,
+    )
+    .context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    for i in 0..=255 {
+        let mut val: usize = 0;
+        write_bytes(&mut val, i, 1);
+        for j in 0..half_count {
+            timeout_checker.check()?;
+            write_volatile(base_ptr.add(j), val);
+            write_volatile(half_ptr.add(j), val);
+        }
+        compare_regions(base_ptr, half_ptr, half_count, timeout_checker)?;
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+// TODO: Consider a more readable format dipslaying MemtestOutcome
+impl fmt::Display for MemtestOutcome {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl fmt::Display for MemtestError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl Error for MemtestError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            MemtestError::Timeout => None,
+            MemtestError::Other(err) => err.source(),
```

Comment:
This says that the source for the `MemtestError` is the source for the `anyhow::Error`, but actually the source for `MemtestError` **is** the `anyhow::Error` itself. Just change this to `MemtestError::Other(err) => Some(err)`